### PR TITLE
Implement MMDIT Single block for Flux

### PR DESCRIPTION
**Pull request by KyleHerndon**

### CODE REVIEW

File: sharktank/sharktank/layers/mmdit.py [hunk 1]
```diff
@@ -144,3 +144,46 @@ def forward(
...
+        )
+        self.add_module(
+            "attn_norm_k", RMSNormLayer(theta("attn.norm.key_norm"), epsilon=1e-6)
+        )
+        self.add_module("attn_proj", LinearLayer(theta("attn.proj")))
+
+        self.add_module("linear1", LinearLayer(theta("linear1")))
+        self.add_module("linear2", LinearLayer(theta("linear2")))
+        self.hidden_size = 3072  # todo
+        self.mlp_hidden_dim = 3072  # todo
```
Comment (sogartar) [inline sharktank/sharktank/layers/mmdit.py:166]: Did you mean to leave these like that?

File: sharktank/sharktank/layers/mmdit.py [hunk 2]
```diff
@@ -144,3 +144,46 @@ def forward(
...
+
+    def forward(self, x: Tensor, vec: Tensor, pe: Tensor) -> tuple[Tensor, Tensor]:
+        mod, _ = self.mod(vec)
+        x_norm = ops.layer_norm(x, None, None, eps=1e-6)
+        x_mod = (1 + mod.scale) * x_norm + mod.shift
+        x_lin = self.linear1(x_mod)
+        print("x_lin_shape", x_lin.shape)
```
Comment (sogartar) [inline sharktank/sharktank/layers/mmdit.py:173]: This print seems like a debugging remanent.

File: sharktank/sharktank/layers/mmdit.py [hunk 3]
```diff
@@ -144,3 +144,46 @@ def forward(
...
+        qkv, mlp = torch.split(
+            x_lin, [3 * self.hidden_size, 4 * self.mlp_hidden_dim], dim=-1
+        )  # todo
+
+        qkv_2 = qkv.view(qkv.shape[0], qkv.shape[1], 3, self.num_heads, -1)  #
+        qkv_3 = ops.permute(qkv_2, (2, 0, 3, 1, 4))
+        q, k, v = qkv_3
+        # q, k, v = rearrange(qkv, "B L (K H D) -> K B H L D", K=3, H=self.num_heads)
```
Comment (sogartar) [inline sharktank/sharktank/layers/mmdit.py:181]: Can you remove this comment?

File: sharktank/sharktank/layers/mmdit.py [hunk 4]
```diff
@@ -144,3 +144,46 @@ def forward(
...
+
+    def forward(self, x: Tensor, vec: Tensor, pe: Tensor) -> tuple[Tensor, Tensor]:
+        mod, _ = self.mod(vec)
+        x_norm = ops.layer_norm(x, None, None, eps=1e-6)
+        x_mod = (1 + mod.scale) * x_norm + mod.shift
+        x_lin = self.linear1(x_mod)
+        print("x_lin_shape", x_lin.shape)
+        qkv, mlp = torch.split(
+            x_lin, [3 * self.hidden_size, 4 * self.mlp_hidden_dim], dim=-1
+        )  # todo
```
Comment (sogartar) [inline sharktank/sharktank/layers/mmdit.py:176]: What is this todo?

File: sharktank/sharktank/layers/testing.py [hunk 5]
```diff
@@ -51,80 +51,139 @@ def make_llama_attention_block_theta(
     )
 
 
-def make_mmdit_double_block_theta(dtype: torch.dtype | None = None) -> Theta:
+def make_mmdit_double_block_theta(
```
Comment (sogartar) [inline sharktank/sharktank/layers/testing.py:54]: The naming would benefit if it clarified that it creates a random theta.

File: sharktank/sharktank/layers/testing.py [hunk 6]
```diff
@@ -51,80 +51,139 @@ def make_llama_attention_block_theta(
...
             ),
             "txt_mod.lin.weight": DefaultPrimitiveTensor(
-                data=make_rand_torch((18432, 3072), dtype=dtype)
+                data=make_rand_torch((mlp_hidden_size3, hidden_size), dtype=dtype)
+            ),
+        }
+    )
+
+
+def make_mmdit_single_block_theta(
```
Comment (sogartar) [inline sharktank/sharktank/layers/testing.py:144]: The naming would benefit if it clarified that it creates a random theta.

**Review by sogartar — APPROVED** [review]

File: sharktank/sharktank/layers/mmdit.py (hunk 1 above)
Comment (KyleHerndon) [reply sharktank/sharktank/layers/mmdit.py]: Fleshed out the TODO, thanks for reminding me.

File: sharktank/sharktank/layers/mmdit.py (hunk 4 above)
Comment (KyleHerndon) [reply sharktank/sharktank/layers/mmdit.py]: Related to the other TODOs, forgot to flesh out.

File: sharktank/sharktank/layers/testing.py (hunk 5 above)
Comment (KyleHerndon) [reply sharktank/sharktank/layers/testing.py]: Done.

File: sharktank/sharktank/layers/testing.py (hunk 6 above)
Comment (KyleHerndon) [reply sharktank/sharktank/layers/testing.py]: Done.

**Pull request merged**